File: WorldFeed/Server/src/WorldFeed.Controllers/Search/SearchQueryExecutor.cs
namespace WorldFeed.Controllers.Search
{
    using System.Threading.Tasks;

    using WorldFeed.Common.DTO;
    using WorldFeed.Common.Public.Models.Enums;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Models.Interfaces.DTO;
    using WorldFeed.Common.Public.Parameters.Search;
    using WorldFeed.Common.Web;

    public interface ISearchQueryExecutor
    {
        Task<ITwitterResult<ISearchResultsDTO>> SearchTweetsAsync(ISearchTweetsParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<UserDTO[]>> SearchUsersAsync(ISearchUsersParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<SavedSearchDTO>> CreateSavedSearchAsync(ICreateSavedSearchParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<SavedSearchDTO>> GetSavedSearchAsync(IGetSavedSearchParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<SavedSearchDTO[]>> ListSavedSearchesAsync(IListSavedSearchesParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<SavedSearchDTO>> DestroySavedSearchAsync(IDestroySavedSearchParameters parameters, ITwitterRequest request);
    }

    public class SearchQueryExecutor : ISearchQueryExecutor
    {
        private readonly ISearchQueryGenerator searchQueryGenerator;
        private readonly ITwitterAccessor twitterAccessor;

        public SearchQueryExecutor(ISearchQueryGenerator searchQueryGenerator, ITwitterAccessor twitterAccessor)
        {
            this.searchQueryGenerator = searchQueryGenerator;
            this.twitterAccessor = twitterAccessor;
        }

        public Task<ITwitterResult<ISearchResultsDTO>> SearchTweetsAsync(ISearchTweetsParameters parameters, ITwitterRequest request)
        {
            request.Query.Url = this.searchQueryGenerator.GetSearchTweetsQuery(parameters, request.ExecutionContext.TweetMode);
            request.Query.HttpMethod = HttpMethod.GET;
            return this.twitterAccessor.ExecuteRequestAsync<ISearchResultsDTO>(request);
        }

        public Task<ITwitterResult<UserDTO[]>> SearchUsersAsync(ISearchUsersParameters parameters, ITwitterRequest request)
        {
            request.Query.Url = this.searchQueryGenerator.GetSearchUsersQuery(parameters);
            request.Query.HttpMethod = HttpMethod.GET;
            return this.twitterAccessor.ExecuteRequestAsync<UserDTO[]>(request);
        }

        public Task<ITwitterResult<SavedSearchDTO>> CreateSavedSearchAsync(ICreateSavedSearchParameters parameters, ITwitterRequest request)
        {
            request.Query.Url = this.searchQueryGenerator.GetCreateSavedSearchQuery(parameters);
            request.Query.HttpMethod = HttpMethod.POST;
            return this.twitterAccessor.ExecuteRequestAsync<SavedSearchDTO>(request);
        }

        public Task<ITwitterResult<SavedSearchDTO>> GetSavedSearchAsync(IGetSavedSearchParameters parameters, ITwitterRequest request)
        {
            request.Query.Url = this.searchQueryGenerator.GetSavedSearchQuery(parameters);
            request.Query.HttpMethod = HttpMethod.GET;
            return this.twitterAccessor.ExecuteRequestAsync<SavedSearchDTO>(request);
        }

        public Task<ITwitterResult<SavedSearchDTO[]>> ListSavedSearchesAsync(IListSavedSearchesParameters parameters, ITwitterRequest request)
        {
            request.Query.Url = this.searchQueryGenerator.GetListSavedSearchQuery(parameters);
            request.Query.HttpMethod = HttpMethod.GET;
            return this.twitterAccessor.ExecuteRequestAsync<SavedSearchDTO[]>(request);
        }

        public Task<ITwitterResult<SavedSearchDTO>> DestroySavedSearchAsync(IDestroySavedSearchParameters parameters, ITwitterRequest request)
        {
            request.Query.Url = this.searchQueryGenerator.GetDestroySavedSearchQuery(parameters);
            request.Query.HttpMethod = HttpMethod.POST;
            return this.twitterAccessor.ExecuteRequestAsync<SavedSearchDTO>(request);
        }
    }
}

File: WorldFeed/Client/src/app/controllers/Search/SearchQueryExecutor.ts
import {ITwitterRequest} from "../../core/Public/Models/Interfaces/ITwitterRequest";
import {ITwitterResult} from "../../core/Core/Web/TwitterResult";
import {ITwitterAccessor} from "../../core/Core/Web/ITwitterAccessor";
import {ComputedTweetMode} from "../../core/Core/QueryGenerators/ComputedTweetMode";
import {HttpMethod} from 'src/app/core/Public/Models/Enum/HttpMethod';
import Task from 'src/app/c#-objects/TypeScript.NET-Core/packages/Threading/source/Tasks/Task';

export interface ISearchQueryExecutor {
  SearchTweetsAsync(parameters: ISearchTweetsParameters, request: ITwitterRequest): Task<ITwitterResult<ISearchResultsDTO>>;

  SearchUsersAsync(parameters: ISearchUsersParameters, request: ITwitterRequest): Task<ITwitterResult<UserDTO[]>>;

  CreateSavedSearchAsync(parameters: ICreateSavedSearchParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO>>;

  GetSavedSearchAsync(parameters: IGetSavedSearchParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO>>;

  ListSavedSearchesAsync(parameters: IListSavedSearchesParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO[]>>;

  DestroySavedSearchAsync(parameters: IDestroySavedSearchParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO>>;
}

export class SearchQueryExecutor implements ISearchQueryExecutor {
  private readonly _searchQueryGenerator: ISearchQueryGenerator;
  private readonly _twitterAccessor: ITwitterAccessor;

  constructor(searchQueryGenerator: ISearchQueryGenerator, twitterAccessor: ITwitterAccessor) {
    this._searchQueryGenerator = searchQueryGenerator;
    this._twitterAccessor = twitterAccessor;
  }

  public SearchTweetsAsync(parameters: ISearchTweetsParameters, request: ITwitterRequest): Task<ITwitterResult<ISearchResultsDTO>> {
    request.query.url = this._searchQueryGenerator.GetSearchTweetsQuery(parameters, new ComputedTweetMode(parameters, request));
    request.query.httpMethod = HttpMethod.GET;
    return this._twitterAccessor.executeRequestAsync<ISearchResultsDTO>(request);
  }

  public SearchUsersAsync(parameters: ISearchUsersParameters, request: ITwitterRequest): Task<ITwitterResult<UserDTO[]>> {
    request.query.url = this._searchQueryGenerator.GetSearchUsersQuery(parameters);
    request.query.httpMethod = HttpMethod.GET;
    return this._twitterAccessor.executeRequestAsync<UserDTO[]>(request);
  }

  public CreateSavedSearchAsync(parameters: ICreateSavedSearchParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO>> {
    request.query.url = this._searchQueryGenerator.GetCreateSavedSearchQuery(parameters);
    request.query.httpMethod = HttpMethod.POST;
    return this._twitterAccessor.executeRequestAsync<SavedSearchDTO>(request);
  }

  public GetSavedSearchAsync(parameters: IGetSavedSearchParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO>> {
    request.query.url = this._searchQueryGenerator.GetSavedSearchQuery(parameters);
    request.query.httpMethod = HttpMethod.GET;
    return this._twitterAccessor.executeRequestAsync<SavedSearchDTO>(request);
  }

  public ListSavedSearchesAsync(parameters: IListSavedSearchesParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO[]>> {
    request.query.url = this._searchQueryGenerator.GetListSavedSearchQuery(parameters);
    request.query.httpMethod = HttpMethod.GET;
    return this._twitterAccessor.executeRequestAsync<SavedSearchDTO[]>(request);
  }

  public DestroySavedSearchAsync(parameters: IDestroySavedSearchParameters, request: ITwitterRequest): Task<ITwitterResult<SavedSearchDTO>> {
    request.query.url = this._searchQueryGenerator.GetDestroySavedSearchQuery(parameters);
    request.query.httpMethod = HttpMethod.POST;
    return this._twitterAccessor.executeRequestAsync<SavedSearchDTO>(request);
  }
}
